### PR TITLE
Check for suspicious gradle-wrapper.jar 

### PR DESCRIPTION
**Pull request by snjeza**
Requires https://github.com/eclipse/eclipse.jdt.ls/pull/1434

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -165,6 +165,7 @@ New in 0.62.0:
   - `autoApply`: Always automatically update the imports and package declarations.
   - `preview`: Always preview the changes before applying.
   - `prompt`: Ask user to confirm whether to bypass refactor preview.
+* `java.imports.gradle.wrapper.checksums`: Defines the allowed gradle wrapper sha256 checksums.
```
Comment (fbricon) [inline README.md:168]: Defines allowed/disallowed SHA-256 checksums of Gradle Wrappers

File: package.json [hunk 2]
```diff
@@ -557,6 +557,32 @@
...
+          "items": {
+            "type": "object",
+            "default": {},
+            "required": [
+              "sha256"
+            ],
+            "properties": {
+              "sha256": {
+                "type": "string",
+                "label": "Sha256 checksum."
```
Comment (fbricon) [inline package.json:572]: SHA-256 checksum

File: package.json [hunk 3]
```diff
@@ -557,6 +557,32 @@
...
+              "allowed": {
+                "type": "boolean",
+                "default": true,
+                "label": "Is allowed?"
+              }
+            },
+            "additionalProperties": false,
+            "uniqueItems": true
+          },
+          "description": "Defines the allowed/disallowed gradle wrapper sha256 checksums.",
```
Comment (fbricon) [inline package.json:583]: Defines allowed/disallowed SHA-256 checksums of Gradle Wrappers

File: src/extension.ts [hunk 4]
```diff
@@ -181,6 +182,7 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 						moveRefactoringSupport: true,
 						clientHoverProvider: true,
 						clientDocumentSymbolProvider: true,
+						actionableNotificationSupported: true,
```
Comment (fbricon) [inline src/extension.ts:185]: gradleChecksumWrapperPromptSupport:true

File: src/extension.ts [hunk 5]
```diff
@@ -132,6 +132,7 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 
 	enableJavadocSymbols();
 
+	const GRADLE_CHECKSUM = "gradle/checksum";
```
Comment (fbricon) [inline src/extension.ts:135]: gradle/checksum/prompt

File: src/extension.ts [hunk 6]
```diff
@@ -336,6 +338,16 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 							_onDidClasspathUpdate.fire(Uri.parse(notification.data));
 							return;
 						}
+						if (notification.message === "__GRADLE_WRAPPER_CHECKSUM__") {
```
Comment (fbricon) [inline src/extension.ts:341]: unnecessary since we already support the gradle/checksum/prompt command

File: src/extension.ts [hunk 7]
```diff
@@ -373,6 +385,14 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 						commands.executeCommand(params.command, ...params.arguments);
 					});
 
+					languageClient.onNotification(GRADLE_CHECKSUM, (handler) => {
```
Comment (fbricon) [inline src/extension.ts:388]: unnecessary since we already support the gradle/checksum/prompt command

File: src/settings.ts [hunk 8]
```diff
@@ -17,6 +17,9 @@ const changeItem = {
 	never: 'Never'
 };
 
+const allow = 'Allow';
```
Comment (fbricon) [inline src/settings.ts:20]: 'Trust'

File: src/settings.ts [hunk 9]
```diff
@@ -17,6 +17,9 @@ const changeItem = {
 	never: 'Never'
 };
 
+const allow = 'Allow';
+const disallow = 'Disallow';
```
Comment (fbricon) [inline src/settings.ts:21]: 'Do not trust'

File: src/settings.ts [hunk 10]
```diff
@@ -204,3 +205,27 @@ export function getJavaServerMode(): ServerMode {
 	return workspace.getConfiguration().get('java.server.launchMode')
 		|| ServerMode.HYBRID;
 }
+
+export function setGradleWrapperChecksum(wrapper: string, sha256?: string) {
+	window.showErrorMessage(`"Security Warning! The gradle wrapper jar '${wrapper}'" with sha256 '${sha256}' can be vulnerable. Do you allow it to be used?";`, disallow, allow)
```
Comment (fbricon) [inline src/settings.ts:210]: `"Security Warning! The gradle wrapper '${wrapper}'" [sha256 '${sha256}'] [could be malicious](https://github.com/redhat-developer/vscode-java/wiki/Gradle-Support#suspicious.wrapper). Should it be trusted?`

File: src/settings.ts [hunk 11]
```diff
@@ -204,3 +205,27 @@ export function getJavaServerMode(): ServerMode {
 	return workspace.getConfiguration().get('java.server.launchMode')
 		|| ServerMode.HYBRID;
 }
+
+export function setGradleWrapperChecksum(wrapper: string, sha256?: string) {
```
Comment (fbricon) [inline src/settings.ts:209]: During startup, the server sends multiple requests to the client, making the dialog flicker 2 or 3 times. There should be a check for the wrapper/checksum combo to see if a dialog is already open, before showing the same message.

That means if we have 2 suspicious files, we should see 2 dialogs.

File: README.md (hunk 1 above)
Comment (snjeza) [reply README.md]: Fixed.

File: package.json (hunk 3 above)
Comment (snjeza) [reply package.json]: Fixed.

File: src/extension.ts (hunk 4 above)
Comment (snjeza) [reply src/extension.ts:185]: Fixed.

File: src/extension.ts (hunk 7 above)
Comment (snjeza) [reply src/extension.ts]: Fixed.

File: src/settings.ts (hunk 8 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

File: src/settings.ts (hunk 10 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

File: src/settings.ts (hunk 9 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

File: src/extension.ts (hunk 6 above)
Comment (snjeza) [reply src/extension.ts]: Fixed.

File: src/extension.ts (hunk 5 above)
Comment (snjeza) [reply src/extension.ts]: Fixed.

File: package.json (hunk 2 above)
Comment (snjeza) [reply package.json]: Fixed.

File: src/settings.ts (hunk 11 above)
Comment (snjeza) [reply src/settings.ts:209]: Fixed.

File: src/settings.ts [hunk 12]
```diff
@@ -204,3 +205,43 @@ export function getJavaServerMode(): ServerMode {
...
+			const entry = property.filter(p => (p.sha256 === sha256));
+			if (entry === null || entry.length === 0) {
+				property.push({ sha256: sha256, allowed: allowed });
+				workspace.getConfiguration().update(key, property, ConfigurationTarget.Global);
+			}
+			removeSha256(sha256);
+		});
+}
+
+function removeSha256(sha256: string) {
```
Comment (fbricon) [inline src/settings.ts:242]: unregisterGradleWrapperPromptDialog

File: src/settings.ts [hunk 13]
```diff
@@ -21,6 +21,7 @@ const EXCLUDE_FILE_CONFIG = 'configuration.checkProjectSettingsExclusions';
 export const ORGANIZE_IMPORTS_ON_PASTE = 'actionsOnPaste.organizeImports'; // java.actionsOnPaste.organizeImports
 
 let oldConfig: WorkspaceConfiguration = getJavaConfiguration();
+const gradlePromptDialogs = [];
```
Comment (fbricon) [inline src/settings.ts:24]: gradleWrapperPromptDialogs

File: src/settings.ts (hunk 12 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

File: src/settings.ts (hunk 13 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

**Comment by snjeza** [comment]
@fbricon I have update the PR.

**Comment by fbricon** [comment]
Before a Gradle wrapper is executed, its checksum is verified against the list of wrappers publicly released by the Gradle team. If the checksum doesn't match a known good checksum, a security warning is displayed (actually an error, so it stays visible):

 Screen Shot 2020-05-20 at 11 29 20 PM
The link opens the [corresponding wiki page](https://github.com/redhat-developer/vscode-java/wiki/Gradle-Support#suspicious.wrapper).

If you trust the wrapper to be executed, the application settings.jon will be updated like:

```json
"java.imports.gradle.wrapper.checksums": [
    {
        "sha256": "504b38a11c466aecb2f5c0b0d8ce0ed7ffa810bf70b9b7a599c570051be8fb4e",
        "allowed": true
    }
],
```
and the wrapper will be used.

If you don't trust it, `"allowed":"false"` will be stored. 

The default Gradle version embedded in Buildship will be used to build your project until the wrapper is explicitly trusted.

**Pull request merged**